### PR TITLE
__CPROVER_r_ok and __CPROVER_w_ok preconditions

### PR DESCRIPTION
**Pull request by kroening**
This adds support for two predicates, which are particularly suitable as preconditions for library models to ensure memory safety of client code.

### CODE REVIEW

**Comment by smowton** [comment]
Guessing this doesn't want a review just yet, please assign when it does

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: ab0d7a9866068c31c78ef3a5e445729a74e80511).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80128022

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -86,6 +86,20 @@ class goto_checkt
...
+      assertion(_assertion),
+      description(_description)
+    {
+    }
+
+    exprt assertion;
+    std::string description;
+  };
+
+  using conditionst=std::list<conditiont>;
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: Spacing

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -1076,6 +1096,109 @@ void goto_checkt::pointer_validity_check(
...
+
+      plus_exprt ub(int_ptr, size, int_ptr.type());
+
+      binary_relation_exprt ub_check(
+        ub, ID_le, plus_exprt(a.first, a.second));
+
+      alloc_disjuncts.push_back(and_exprt(lb_check, ub_check));
+    }
+
+    const exprt allocs=disjunction(alloc_disjuncts);
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: Spacing

File: src/analyses/goto_check.cpp [hunk 3]
```diff
@@ -1076,6 +1096,109 @@ void goto_checkt::pointer_validity_check(
...
+
+    if(flags.is_unknown() || flags.is_null())
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: Prefer braces round multi-line `if` (here and below)

File: src/analyses/goto_check.cpp [hunk 4]
```diff
@@ -1076,6 +1096,109 @@ void goto_checkt::pointer_validity_check(
...
+      conditions.push_back(conditiont(
+        or_exprt(allocs, not_exprt(null_pointer(address))),
```
Comment (smowton) [inline src/analyses/goto_check.cpp]: Surely "assert x in some alloc || x != null" is redundant, unless we want to handle weird systems where you can map the zero page?

**Review by smowton — COMMENTED** [review]
Couple of style nitpicks, and a proper concern: the conditions imposed appear to be redundant?

File: src/analyses/goto_check.cpp (hunk 4 above)
Comment (peterschrammel) [reply src/analyses/goto_check.cpp]: Does `allocs` include mmio?

Comment (kroening) [reply src/analyses/goto_check.cpp]: Yes, I did think about this. Yes, we may wish to be able to do MMIO on address zero.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 323859ae5eca962aaf97b1f0e967d5d3d81e480d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80214792
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b07710aa6f3162b69a7cf5cb1d59ac73cc864cbd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80216261

File: src/ansi-c/c_typecheck_expr.cpp [hunk 5]
```diff
@@ -2334,6 +2334,23 @@ exprt c_typecheck_baset::do_special_functions(
 
     return malloc_expr;
   }
+  else if(identifier==CPROVER_PREFIX "r_ok" ||
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_expr.cpp]: Nit pick: clang-format style spacing (across this commit)

File: src/util/irep_ids.def [hunk 6]
```diff
@@ -691,6 +691,8 @@ IREP_ID_TWO(C_unnamed_object, #unnamed_object)
 IREP_ID_TWO(C_temporary_avoided, #temporary_avoided)
 IREP_ID_TWO(C_qualifier, #qualifier)
 IREP_ID_TWO(C_array_ini, #array_ini)
+IREP_ID_ONE(r_ok)
```
Comment (tautschnig) [inline src/util/irep_ids.def:694]: Randomly placing this here: this badly needs comments/documentation: the commit message has zero information, and RTFS doesn't help either as the front-end just quietly eats the functions but no semantics are ever provided.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 5 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_expr.cpp]: done

File: src/util/irep_ids.def (hunk 6 above)
Comment (kroening) [reply src/util/irep_ids.def:694]: done

**Comment by tautschnig** [comment]
I'm ok with this in principle, but it seems that at least one of the commits introduces regressions on selected tests. Once that's debugged it should be good to go.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: ae10d175e26a0269e5341efac95d4e8183e91537).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80409979

**Comment by kroening** [comment]
The AWS CodeBuild windows failure is spurious (this time it's the choco server)

File: regression/cbmc/Function5/test.desc [hunk 7]
```diff
@@ -3,6 +3,6 @@ main.c
 --pointer-check --bounds-check
 ^SIGNAL=0$
 ^EXIT=10$
-^\[.*\] dereference failure: pointer outside object bounds in \*p: FAILURE$
+^\[.*\] pointer outside object bounds in \*p: FAILURE$
```
Comment (tautschnig) [inline regression/cbmc/Function5/test.desc]: I'm quite surprised to read a commit message saying "refactor ..." and then see user-visible output change. What is the rationale for this change? Could it please be communicated (at bare minimum via the commit message) as this will impact users? (For a start, the SV-COMP scripts will need to be updated.)

Comment (kroening) [reply regression/cbmc/Function5/test.desc]: Ok, removed change to message

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 4b5a7f0e9126c12ee74db37ba556346ab1610b4e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80421956

File: src/analyses/goto_check.cpp [hunk 8]
```diff
@@ -1418,24 +1392,38 @@ void goto_checkt::check_rec(const exprt &expr, guardt &guard, bool address)
...
-    if(member_offset.is_not_nil() && size.is_not_nil())
-      access_ub=plus_exprt(member_offset, size);
+    if(member_offset.is_not_nil())
+    {
+      pointer_typet new_pointer_type = to_pointer_type(deref.pointer().type());
+      new_pointer_type.subtype() = expr.type();
 
-    pointer_validity_check(deref, guard, member_offset, access_ub);
+      const exprt char_pointer =
+        typecast_exprt::conditional_cast(deref.pointer(), pointer_type(char_type()));
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: cpp-lint says this line is too long (82 chars).

**Review by tautschnig — APPROVED** [review]
There are spurious failures in the Travis job, but adjusting the line length (see below) will trigger a re-run anyway.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0202f34620e27bc3a679b0c720e978d7afe75d99).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80458659

**Review by smowton — APPROVED** [review]
